### ISSUE TITLE
shap_interaction_values rows not adding up to shap values for RF classifier

### ISSUE THREAD

**Issue by oegedijk** [shap/shap#723]
Hi all,

I have some issues with getting shap_interaction_values to work with a classification problem. To use the example code in [this tutorial](https://github.com/slundberg/shap/blob/master/notebooks/tree_explainer/Basic%20SHAP%20Interaction%20Value%20Example%20in%20XGBoost.ipynb) to generate X and y:

 If you then fit an sklearn RandomForestRegressor it indeed works as advertised:

```
from sklearn.ensemble import RandomForestRegressor
rfr = RandomForestRegressor()
rfr.fit(X, y)
reg_preds = rfr.predict(X)

reg_explainer = shap.TreeExplainer(rfr)
reg_shap_interact_vals = reg_explainer.shap_interaction_values(X)
reg_shap_vals = reg_explainer.shap_values(X)

reg_shap_vals[0].round(2)
reg_shap_interact_vals[0].round(2)
reg_shap_interact_vals[0].sum(0).round(2)

np.abs(reg_shap_vals.sum(1) + reg_explainer.expected_value - reg_preds).max()
```
Output for the first row (index 0):

```
array([ 1.  , -1.25,  0.25,  0.  ,  0.  ])
array([[ 1.  , -0.01, -0.  ,  0.  ,  0.  ],
       [-0.01, -1.5 ,  0.25,  0.  ,  0.  ],
       [-0.  ,  0.25, -0.  ,  0.  ,  0.  ],
       [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ],
       [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ]])
array([ 1.  , -1.25,  0.25,  0.  ,  0.  ])
6.124658247230741e-08
```

However when we turn it into a classification problem by generating random 1 and 0 depending on the value of the original y's:
```
y_prob =  1 / (1 + np.exp(-y)) #sigmoid
y_cls = np.where(np.random.rand(2000) > 1-y_prob, 1, 0)

from sklearn.ensemble import  RandomForestClassifier
rfc = RandomForestClassifier()
rfc.fit(X, y_cls)
cls_preds = rfc.predict_proba(X)

cls_explainer = shap.TreeExplainer(rfc)
cls_shap_interact_vals = cls_explainer.shap_interaction_values(X)[1]
cls_shap_vals = cls_explainer.shap_values(X)[1]
```

Then the interaction values no longer sum up to the shap values, and what's more the values on the diagonals (which should be the main effect) are simply equal to the shap_values themselves.

```
cls_shap_vals[0].round(2)
cls_shap_interact_vals[0].round(2)
cls_shap_interact_vals[0].sum(0).round(2)

np.abs(cls_shap_vals.sum(1) + cls_explainer.expected_value[1]- cls_preds[:,1]).max()
np.abs(cls_shap_interact_vals.sum(2).sum(1) + cls_explainer.expected_value[1] - cls_preds[:,1]).max()
```

Output for the first row (index 0):
```
array([ 0.17, -0.24,  0.07, -0.01,  0.  ])
array([[ 0.17,  0.01,  0.02, -0.  ,  0.  ],
       [ 0.01, -0.24,  0.04,  0.  ,  0.  ],
       [ 0.02,  0.04,  0.07, -0.  ,  0.  ],
       [-0.  ,  0.  , -0.  , -0.01,  0.  ],
       [ 0.  ,  0.  ,  0.  ,  0.  ,  0.  ]])
array([ 0.2 , -0.19,  0.13, -0.01,  0.  ])
1.0566299279979319e-08
0.14718125134184412
```

So the row sums of the interaction values sum up to `array([ 0.2 , -0.19,  0.13, -0.01,  0.  ])` while the shap values are `array([ 0.17, -0.24,  0.07, -0.01,  0.  ])`. 

However the values on the diagonal are exactly equal to `array([ 0.17, -0.24,  0.07, -0.01,  0.  ])`. Which begs the question: what are the values off the diagonal then?

Should interaction values for a classification problem be generated differently? Or is the algorithm malfunctioning somewhere?

**Comment by oegedijk** [comment]
So I wrote a workaround that fixes the row sums to add up to the shap_values:

```
def normalize_shap_interaction_values(shap_interaction_values, shap_values=None):
    """
    Normalizes shap_interactiong_values to make sure that the rows add up to the shap_values.
    
    This is a workaround for an apparant bug where the diagonals of shap_interaction_values 
    of a RandomForestClassifier are set equal to the shap_values instead of the main effect. 
    """
    siv = shap_interaction_values.copy()
    
    orig_diags = np.einsum('ijj->ij', siv)
    row_sums = np.einsum('ijk->ij', siv)
    row_diffs = row_sums - orig_diags # sum of rows excluding diagonal elements
    
    if shap_values is not None:
        diags = shap_values - row_diffs
    else:
        # if no shap_values provided assume that the original diagonal values were indeed equal to
        # the shap values, and so simply 
        diags = orig_diags - row_diffs
    
    s0,s1,s2 = siv.shape
    siv.reshape(s0,-1)[:,::s2+1] = diags
    return siv
```

If you normalize the `shap_interaction_values`, the rows no do add up to the shap_values:
```
cls_shap_interact_vals2 = normalize_shap_interaction_values(cls_shap_interact_vals, cls_shap_vals)
```

```
cls_shap_vals[0]
cls_shap_interact_vals2[0]
cls_shap_interact_vals2[0].sum(0)
np.abs(cls_shap_vals.sum(1) + cls_explainer.expected_value[1]- cls_preds[:,1]).max()
np.abs(cls_shap_interact_vals2.sum(2).sum(1) + cls_explainer.expected_value[1] - cls_preds[:,1]).max()
np.abs(cls_shap_interact_vals2.sum(2).sum(1) - cls_shap_vals.sum(1)).max()
```

Output:

```
array([ 0.16847716, -0.225861  ,  0.06925555,  0.00427386,  0.        ])
array([[ 0.14725953, -0.00517753,  0.02217857,  0.0042166 ,  0.        ],
       [-0.00517753, -0.25840557,  0.03876833, -0.00104623,  0.        ],
       [ 0.02217857,  0.03876833,  0.00421705,  0.0040916 ,  0.        ],
       [ 0.0042166 , -0.00104623,  0.0040916 , -0.00298812,  0.        ],
       [ 0.        ,  0.        ,  0.        ,  0.        ,  0.        ]])
array([ 0.16847716, -0.225861  ,  0.06925555,  0.00427386,  0.        ])
1.309327668508331e-08
1.309327668508331e-08
5.551115123125783e-17
```

When plotting the main effects with the original shap_interaction_values we see dispersal (which shouldn't happen for the main effect):
`shap.dependence_plot((0,0), cls_shap_interact_vals, X)`:

![image](https://user-images.githubusercontent.com/27999937/62122475-12c9e200-b2c6-11e9-8bbd-c96b5426a889.png)

Whereas with the normalized it looks better, although there is still a bit of jitter (numerical rounding errors?):

`shap.dependence_plot((0,0), normalize_shap_interaction_values(cls_shap_interact_vals, cls_shap_vals), X)`:

![image](https://user-images.githubusercontent.com/27999937/62122595-59b7d780-b2c6-11e9-9411-777674cc972b.png)

Haven't looked at what caused the underlying apparant bug though...
